File: yua-sdk/src/core/api-client.ts
import { APIError } from "./error";

export interface APIClientConfig {
  baseURL: string;
  apiKey?: string;
  authProvider?: () => Promise<string>;
  workspace?: string;
  timeout: number;
  maxRetries: number;
}

export class APIClient {
  private readonly config: APIClientConfig;

  constructor(config: APIClientConfig) {
    this.config = {
      ...config,
      baseURL: config.baseURL.replace(/\/+$/, ""),
    };
  }

  private async buildHeaders(): Promise<Record<string, string>> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json",
    };

    if (this.config.apiKey) {
      // Backend checks x-api-key header, hashes with SHA256, looks up api_keys_v2
      headers["x-api-key"] = this.config.apiKey;
    } else if (this.config.authProvider) {
      // Firebase token → Authorization: Bearer
      const token = await this.config.authProvider();
      headers["Authorization"] = `Bearer ${token}`;
    }

    if (this.config.workspace) {
      headers["x-workspace-id"] = this.config.workspace;
    }

    return headers;
  }

  private resolveURL(path: string): string {
    const normalized = path.startsWith("/") ? path : `/${path}`;
    return `${this.config.baseURL}${normalized}`;
  }

  async get<T = unknown>(path: string): Promise<T> {
    return this.request("GET", path);
  }

  async post<T = unknown>(path: string, body?: unknown): Promise<T> {
    return this.request("POST", path, body);
  }

  async put<T = unknown>(path: string, body?: unknown): Promise<T> {
    return this.request("PUT", path, body);
  }

  async patch<T = unknown>(path: string, body?: unknown): Promise<T> {
    return this.request("PATCH", path, body);
  }

  async delete<T = unknown>(path: string): Promise<T> {
    return this.request("DELETE", path);
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    attempt = 0,
  ): Promise<T> {
    const headers = await this.buildHeaders();
    const controller = new AbortController();
    const timer = setTimeout(
      () => controller.abort(),
      this.config.timeout,
    );

    try {
      const res = await fetch(this.resolveURL(path), {
        method,
        headers,
        body: body != null ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      if (!res.ok) {
        const error = await APIError.fromResponse(res);

        if (error.isRetryable && attempt < this.config.maxRetries) {
          const delay = Math.min(1000 * 2 ** attempt, 8000);
          await new Promise((r) => setTimeout(r, delay));
          return this.request(method, path, body, attempt + 1);
        }

        throw error;
      }

      return (await res.json()) as T;
    } finally {
      clearTimeout(timer);
    }
  }

  async fetchSSE(
    path: string,
    signal?: AbortSignal,
  ): Promise<ReadableStream<Uint8Array>> {
    const headers = await this.buildHeaders();
    headers["Accept"] = "text/event-stream";
    delete headers["Content-Type"];

    const res = await fetch(this.resolveURL(path), {
      method: "GET",
      headers,
      signal,
    });

    if (!res.ok) {
      throw await APIError.fromResponse(res);
    }

    if (!res.body) {
      throw new APIError(500, "no_body", "SSE response has no body");
    }

    return res.body;
  }
}

File: yua-sdk-python/src/yua/_core/api_client.py
from __future__ import annotations

import time
from typing import Any, Callable, Dict, Optional

import httpx

from yua._errors import APIError

DEFAULT_BASE_URL = "https://api.yuaone.com"
DEFAULT_TIMEOUT = 30.0
DEFAULT_MAX_RETRIES = 2


class APIClient:
    def __init__(
        self,
        *,
        base_url: str = DEFAULT_BASE_URL,
        api_key: Optional[str] = None,
        auth_provider: Optional[Callable[[], str]] = None,
        workspace: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._auth_provider = auth_provider
        self._workspace = workspace
        self._timeout = timeout
        self._max_retries = max_retries
        self._http = httpx.Client(timeout=timeout)

    def close(self) -> None:
        self._http.close()

    def __enter__(self) -> "APIClient":
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    def _build_headers(self) -> Dict[str, str]:
        headers: Dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if self._api_key:
            headers["x-api-key"] = self._api_key
        elif self._auth_provider:
            token = self._auth_provider()
            headers["Authorization"] = f"Bearer {token}"
        if self._workspace:
            headers["x-workspace-id"] = self._workspace
        return headers

    def _url(self, path: str) -> str:
        if not path.startswith("/"):
            path = f"/{path}"
        return f"{self._base_url}{path}"

    def get(self, path: str) -> Any:
        return self._request("GET", path)

    def post(self, path: str, body: Optional[Any] = None) -> Any:
        return self._request("POST", path, body)

    def put(self, path: str, body: Optional[Any] = None) -> Any:
        return self._request("PUT", path, body)

    def patch(self, path: str, body: Optional[Any] = None) -> Any:
        return self._request("PATCH", path, body)

    def delete(self, path: str) -> Any:
        return self._request("DELETE", path)

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[Any] = None,
        attempt: int = 0,
    ) -> Any:
        headers = self._build_headers()
        kwargs: Dict[str, Any] = {"headers": headers}
        if body is not None:
            kwargs["json"] = body

        response = self._http.request(method, self._url(path), **kwargs)

        if not response.is_success:
            error = APIError.from_response(response)
            if error.is_retryable and attempt < self._max_retries:
                delay = min(1.0 * (2**attempt), 8.0)
                time.sleep(delay)
                return self._request(method, path, body, attempt + 1)
            raise error

        return response.json()

    def stream_sse(self, path: str) -> httpx.Response:
        headers = self._build_headers()
        headers["Accept"] = "text/event-stream"
        del headers["Content-Type"]

        response = self._http.send(
            self._http.build_request("GET", self._url(path), headers=headers),
            stream=True,
        )

        if not response.is_success:
            response.read()
            raise APIError.from_response(response)

        return response
